src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +621,205 @@ public static string GetTempPath(string workingDir)
+                OwnsMutex = true;
+            }
+        }
+
+        public bool TryLock (int timeoutMs)
+        {
+            if (IsLocked)
+                throw new InvalidOperationException("Lock already held");
+
+            var sw = Stopwatch.StartNew();
+            do
+            {
+                try
+                {
+                    Stream.Lock(0, 0);
+                    IsLocked = true;
+                    return true;
+                }
+                catch (IOException)
+                {
+                    // Lock currently held by someone else
+                    Thread.Sleep(0);
+                }
+                catch (Exception)
+                {
+                    // Something else went wrong.
+                    return false;
+                }
+            } while (sw.ElapsedMilliseconds < timeoutMs);
+
+            return false;
+        }
+
+        public void Unlock ()
+        {
+            if (!IsLocked)
+                return;
+            Stream.Unlock(0, 0);
+            IsLocked = false;
+        }
+
+        public void Dispose ()
+        {
+            var wasLocked = IsLocked;
+            if (wasLocked)
+                Unlock();
+            Stream.Dispose();
+            // We do not delete the lock file here because there is no reliable way to perform a
+            //  'delete if no one has the file open' operation atomically on *nix. This is a leak.
+        }
+    }
+
+    internal sealed class ServerNamedMutex : IServerMutex
+    {
+        private readonly Mutex Mutex;
+
+        public bool IsDisposed { get; private set; }
+        public bool IsLocked { get; private set; }
+
+        public ServerNamedMutex (string mutexName, bool createNewLocked, out bool createdNew)
+        {
+            if (createNewLocked) {
+                Mutex = new Mutex(
+                    initiallyOwned: true,
+                    name: mutexName,
+                    createdNew: out createdNew
+                );
+                if (createdNew)
+                    IsLocked = true;
+            } else {
+                Mutex.TryOpenExisting(mutexName, out Mutex);
+                createdNew = false;
+            }
+
+            IsDisposed = (Mutex == null);
+        }
+
+        public bool TryLock (int timeoutMs)
+        {
+            if (IsDisposed)
+                return false;
+            if (IsLocked)
+                throw new InvalidOperationException("Lock already held");
+            return IsLocked = Mutex.WaitOne(timeoutMs);
+        }
+
+        public void Unlock ()
+        {
+            if (IsDisposed)
+                return;
+            if (!IsLocked)
+                return;
+            Mutex.ReleaseMutex();
+            IsLocked = false;
+        }
+
+        public void Dispose ()
+        {
+            if (IsDisposed)
+                return;
+            IsDisposed = true;
+
+            if (IsLocked)
+                Mutex?.ReleaseMutex();
```

Comment:
`ReleaseMutex` can throw several exceptions which would cause the remainder of the `Dispose` to be skipped. Lets use a `try / finally` here and put pretty much all the logic there. 

```
try
{
    if (IsLocked)
        Mutex?.ReleaseMutex();
}
finally
{
    IsDisposed = true;
    IsLocked = false;
    Mutex?.Dispose();
}

```